File: backend/app/utils/url_validator.py
"""
Strict URL Validation Utility for PhishGuard Threat Analysis Engine.
"""

import re
from urllib.parse import urlparse
from typing import Tuple


def validate_strict_url(target: str) -> Tuple[bool, str]:
    """
    Validates that target is strictly a valid HTTP/HTTPS URL.
    Returns (is_valid, error_message).
    """
    if not target or not isinstance(target, str):
        return False, "It is not a valid URL"

    trimmed = target.strip()
    if not trimmed:
        return False, "It is not a valid URL"

    # Must start with http:// or https://
    if not re.match(r"^https?://", trimmed, re.IGNORECASE):
        return False, "It is not a valid URL"

    # Reject internal spaces
    if any(c.isspace() for c in trimmed):
        return False, "It is not a valid URL"

    # Reject prohibited schemes
    if re.match(r"^(javascript|data|file|ftp|mailto|about|blob|vbscript):", trimmed, re.IGNORECASE):
        return False, "It is not a valid URL"

    try:
        parsed = urlparse(trimmed)

        # Protocol must be http or https
        if parsed.scheme.lower() not in ("http", "https"):
            return False, "It is not a valid URL"

        hostname = parsed.hostname
        if not hostname:
            return False, "It is not a valid URL"

        # Hostname cannot contain spaces or invalid control characters
        if re.search(r"[\s\r\n\t]", hostname):
            return False, "It is not a valid URL"

        # Cannot be plain single number (e.g. http://123456789)
        if hostname.isdigit():
            return False, "It is not a valid URL"

        # Check IPv4
        ipv4_match = re.match(r"^(\d{1,3}\.){3}\d{1,3}$", hostname)
        if ipv4_match:
            parts = [int(p) for p in hostname.split(".")]
            if all(0 <= p <= 255 for p in parts):
                return True, ""
            return False, "It is not a valid URL"

        # Check localhost or bracketed IPv6
        if hostname == "localhost" or (hostname.startswith("[") and hostname.endswith("]")):
            return True, ""

        # Standard domain must contain at least one dot
        if "." not in hostname:
            return False, "It is not a valid URL"

        domain_parts = hostname.split(".")
        tld = domain_parts[-1]

        # TLD must be at least 2 chars and not all numeric
        if len(tld) < 2 or tld.isdigit() or not re.match(r"^[a-zA-Z0-9-]+$", tld):
            return False, "It is not a valid URL"

        for part in domain_parts:
            if not part or part.startswith("-") or part.endsWith("-") if hasattr(part, "endsWith") else part.endswith("-"):
                return False, "It is not a valid URL"
            if not re.match(r"^[a-zA-Z0-9-]+$", part):
                return False, "It is not a valid URL"

        return True, ""
    except Exception:
        return False, "It is not a valid URL"

File: frontend/src/utils/urlValidator.js
/**
 * Strict URL Validation Utility for PhishGuard Threat Scanner
 *
 * Rules:
 * 1. Must be a non-empty string with no internal unencoded spaces.
 * 2. Must start with http:// or https:// (case-insensitive).
 * 3. Must have a valid hostname (domain with TLD or valid IP / localhost).
 * 4. Must reject plain text, numbers, emails, usernames, unsupported schemes (javascript:, file:, ftp:, mailto:),
 *    and domain-like strings without protocol (e.g. google.com).
 */

export function isValidUrl(input) {
  if (!input || typeof input !== 'string') {
    return false;
  }

  const trimmed = input.trim();
  if (!trimmed) {
    return false;
  }

  // 1. Must strictly start with http:// or https://
  const schemeRegex = /^https?:\/\//i;
  if (!schemeRegex.test(trimmed)) {
    return false;
  }

  // 2. Must not contain unencoded spaces or line breaks
  if (/\s/.test(trimmed)) {
    return false;
  }

  // 3. Reject prohibited schemes or script injection patterns
  const prohibitedSchemes = /^(javascript|data|file|ftp|mailto|about|blob|vbscript):/i;
  if (prohibitedSchemes.test(trimmed)) {
    return false;
  }

  // 4. Parse using URL constructor
  try {
    const parsed = new URL(trimmed);

    // Protocol must strictly be http: or https:
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
      return false;
    }

    const hostname = parsed.hostname;
    if (!hostname || hostname.length === 0) {
      return false;
    }

    // Hostname cannot contain spaces or invalid characters
    if (/[\s\r\n\t]/.test(hostname)) {
      return false;
    }

    // Hostname cannot be purely numeric without dots (e.g. http://123456789)
    if (/^\d+$/.test(hostname)) {
      return false;
    }

    // Check if hostname is an IPv4 address
    const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
    if (ipv4Regex.test(hostname)) {
      const parts = hostname.split('.').map(Number);
      const validIpv4 = parts.every((p) => p >= 0 && p <= 255);
      if (!validIpv4) return false;
      return true;
    }

    // Check if hostname is localhost or IPv6
    if (hostname === 'localhost' || hostname.startsWith('[') && hostname.endsWith(']')) {
      return true;
    }

    // For standard domains, must contain at least one dot and a valid TLD
    // Reject plain strings like http://google or http://test
    if (!hostname.includes('.')) {
      return false;
    }

    // Domain name parts check
    const domainParts = hostname.split('.');
    const tld = domainParts[domainParts.length - 1];

    // TLD must be at least 2 characters and only letters (or punycode xn--)
    if (!tld || tld.length < 2 || !/^[a-zA-Z0-9-]+$/.test(tld) || /^\d+$/.test(tld)) {
      return false;
    }

    // Each label in domain must be non-empty and valid
    for (const part of domainParts) {
      if (!part || part.length === 0 || part.startsWith('-') || part.endsWith('-')) {
        return false;
      }
      if (!/^[a-zA-Z0-9-]+$/.test(part)) {
        return false;
      }
    }

    return true;
  } catch (err) {
    return false;
  }
}
